rfcs/2020-11-02-remap-strict-return-value-types.md

```diff
@@ -0,0 +1,278 @@
+a check against an event, and return a boolean outcome. For example, in the
+`swimlanes` transform, the check determines if an event should be included in a
+specific swimlane.
+
+Currently, any non-boolean return value of a remap program results in a failure
+of the condition, which can be confusing to users.
+
+## Terminology
+
+- `source` — The Remap source code written by users, to compile and execute.
+- `program` — A list of expressions, compiled from source and a list of
+  functions, ready to run against an object.
+- `expression` — A piece of source code that resolves to a value.
+- `value` — a concrete value type, such as string, integer, boolean, etc.
+- `object` — A generic term for whatever outside data container the program
+  manipulates (such as an "event").
+
+## Internal Proposal
+
+Remap is an expression-based language. Everything, including statements, is an
+expression. Meaning, everything can return a value.
+
+**expression examples**
+
+For example, the program `.foo = "bar"` consists of two expressions:
+
+1. The literal expression `"bar"`, which resolves to a `Value::String`.
+2. The assignment expression `.foo = …`, which resolves to the value of the
+   expression on the right side of the assignment operator.
+
+This example program, when executed, always resolves to a `String` value.
+
+Another example program is `floor(.bar)`. This program again consists of two
+expressions:
+
+1. The query path expression `.bar`, which resolves to any value the `bar`
+   object field contains.
+2. The function expression `floor`, which resolves to either an `Integer`, or
+   `Float`.
+
+Since the `floor` function is the final expression that runs in the program, the
+outcome of a _successful execution_ of the program is either an `Integer` or
+`Float`.
+
+**expression trait**
+
+Expressions are implemented using the `Expression` trait:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+}
+```
+
+Expressions can be fallible (not covered in this RFC), can return `None` (not
+covered in-depth in this RFC), or return a resolved `Value`.
+
+**defining the expected resolved values**
+
+To allow the execution context of a program to determine the outcome of running
+the program, we extend the `Expression` trait as follows:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+
+    /// Describes which [`ValueKind`]s the expression can resolve to.
+    ///
+    /// bike-shedding of function and enum names welcome.
+    fn resolves_to(&self) -> ResolveKind;
+}
+```
+
+We define `ResolveKind` as follows:
+
+```rust
+pub enum ResolveKind {
+    /// The expression can resolve to any value at runtime.
+    Any,
+
+    /// The expressions resolves to one of the defined [`ValueKind`]s.
+    OneOf(Vec<ValueKind>),
```

Comment:
I forgot to expand on what `ValueKind` means. I'll write this up later, but it's basically and enum of the kinds of values the `Value` enum can contain (string, integer, float, etc).

It should also be noted that this only allows for checking "primitive" values, as in, you can check that the value kind is a string, integer, etc, and you can also check if it's a map or array, but you can't define what value kinds are included in the map or array.

In other words, you can't define the expectation of nested value kinds, only the top-level value kind is considered.